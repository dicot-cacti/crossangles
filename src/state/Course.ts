import { StreamData } from './Stream';
import { MinistryMeta } from './Meta';
import { Colour } from './Colours';

export type CourseId = string;

export interface CourseData {
  code: string,
  name: string,
  streams: StreamData[],
  term?: string | null,
  isCustom?: boolean,
  isAdditional?: boolean,
  autoSelect?: boolean,
  defaultColour?: Colour,
  description?: string,
  metadata?: MinistryMeta,
}

export interface CourseMap {
  [id: string]: CourseData,
}


export const getCourseId = (course: CourseData): CourseId => {
  return course.code + (course.term || '');
}

export const hasWebStream = (course: CourseData): boolean => {
  return getWebStream(course) !== null;
}

export const getWebStream = (course: CourseData): StreamData | null => {
  const streams = course.streams;
  for (let i = 0; i < streams.length; ++i) {
    if (streams[i].web) {
      return streams[i];
    }
  }

  return null;
}

export const courseSort = (a: CourseData, b: CourseData) => +(a.code > b.code) - +(a.code < b.code);
export const customSort = (a: CourseData, b: CourseData) => +(a.name > b.name) - +(a.name < b.name);
