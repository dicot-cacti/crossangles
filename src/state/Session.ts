import { CourseData, CourseId, getCourseId } from './Course';
import { StreamData, StreamId, getStreamId, LinkedStream } from './Stream';

export type SessionId = string;

export type DayLetter = 'M' | 'T' | 'W' | 'H' | 'F';

export interface SessionCommon {
  index: number,
  start: number,
  end: number,
  day: DayLetter,
  canClash?: boolean,
  location?: string,
  weeks?: string,
}

export interface SessionData extends SessionCommon {
  course: CourseId,
  stream: StreamId,
}

export interface LinkedSession extends SessionCommon {
  course: CourseData,
  stream: LinkedStream,
  id: SessionId,
}


export const getSessionId = (course: CourseData, stream: StreamData, session: SessionData) => {
  const streamId = getStreamId(course, stream);
  return `${streamId}~${session.index}`;
}

export const linkSession = (course: CourseData, stream: LinkedStream, session: SessionData): LinkedSession => {
  return {
    ...session,
    course,
    stream,
    id: getSessionId(course, stream, session),
  };
}

export const unlinkSession = (session: LinkedSession): SessionData => {
  return {
    index: session.index,
    start: session.start,
    end: session.end,
    day: session.day,
    canClash: session.canClash,
    location: session.location,
    weeks: session.weeks,
    course: getCourseId(session.course),
    stream: getStreamId(session.course, session.stream),
  };
}
