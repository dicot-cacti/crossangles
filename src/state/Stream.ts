import { SessionData, DayLetter, LinkedSession, linkSession } from './Session';
import { CourseData, getCourseId } from './Course';

export type StreamId = string;

export interface StreamData {
  component: string,
  enrols: [number, number],
  times: ClassTime[] | null,
  full?: boolean,
  web?: boolean,
}

export interface LinkedStream extends StreamData {
  course: CourseData,
  sessions: LinkedSession[],
  id: StreamId,
}

export interface ClassTime {
  time: string,
  location?: string,
  weeks?: string,
  canClash?: boolean,
}


export const getStreamId = (course: CourseData, stream: StreamData) => {
  const timeString = stream.times ? stream.times.map(t => t.time).join(',') : 'WEB';
  const id = `${getComponentId(course, stream)}~${timeString}`;
  return id;
}

export const getComponentId = (course: CourseData, stream: StreamData) => {
  const componentString = course.isCustom ? '' : stream.component;
  return `${getCourseId(course)}~${componentString}`;
}

export const getSessions = (course: CourseData, stream: StreamData): SessionData[] => {
  const courseId = getCourseId(course);
  const streamId = getStreamId(course, stream);
  if (stream.times !== null) {
    return stream.times.map((t, i): SessionData => {
      const [ startHour, endHour ] = t.time.substr(1).split('-').map(x => parseFloat(x));
      return {
        start: startHour,
        end: endHour || (startHour + 1),
        day: t.time.charAt(0) as DayLetter,
        canClash: t.canClash,
        location: t.location,
        index: i,
        weeks: t.weeks,
        stream: streamId,
        course: courseId,
      };
    });
  } else {
    return [];
  }
}

export const linkStream = (course: CourseData, stream: StreamData): LinkedStream => {
  const sessionData = getSessions(course, stream);
  const linkedStream: LinkedStream = {
    ...stream,
    course,
    sessions: [],
    id: getStreamId(course, stream),
  };
  linkedStream.sessions = sessionData.map(session => linkSession(course, linkedStream, session));
  return linkedStream;
}
