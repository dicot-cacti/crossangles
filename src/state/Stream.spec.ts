import { getStreamId, getSessions, StreamData, linkStream, getComponentName, closestMonday, parseBackwardsDateString } from './Stream';
import { getCourse } from '../test_util';
import { CourseData, getCourseId } from './Course';
import { SessionData } from './Session';


describe('getStreamId', () => {
  it('returns expected value', () => {
    const course = getCourse();
    const stream = course.streams[0];
    expect(getStreamId(course, stream)).toBe('RING9731~LEC~M9,H12,T12');
  });
});

describe('getSessions', () => {
  it.each([
    true, false,
  ])('returns empty list for stream with no times (web=$web)', web => {
    const stream: StreamData = { component: 'LEC', enrols: [0, 0], times: [], web };
    const course: CourseData = {
      ...getCourse(),
      streams: [stream],
    };
    expect(getSessions(course, stream)).toEqual([]);
  });

  it('gives expected result', () => {
    const stream: StreamData = {
      component: 'LEC',
      enrols: [0, 0],
      times: [{ time: 'M10-20', location: 'foo' }, { time: 'T15', weeks: 'bar' }, { time: 'H8' }],
    };
    const course: CourseData = {
      ...getCourse(),
      streams: [stream],
    };
    const courseId = getCourseId(course);
    const streamId = getStreamId(course, stream);
    const common: Omit<SessionData, 'index' | 'day' | 'start' | 'end'> = {
      course: courseId,
      stream: streamId,
      canClash: undefined,
      weeks: undefined,
      location: undefined,
    };
    const expected: SessionData[] = [
      { ...common, index: 0, day: 'M', start: 10, end: 20, location: 'foo' },
      { ...common, index: 1, day: 'T', start: 15, end: 16, weeks: 'bar' },
      { ...common, index: 2, day: 'H', start: 8, end: 9 },
    ];
    expect(getSessions(course, stream)).toEqual(expected);
  });
});


describe('linkStream', () => {
  it('gives consistent output', () => {
    const course = getCourse();
    for (const stream of course.streams) {
      expect(linkStream(course, stream)).toMatchSnapshot();
    }
  });
});


it.each([
  ['TUT', 'Tutorial'],
  ['LEC', 'Lecture'],
  ['LE1', 'Lecture (1)'],
  ['QQQ', 'QQQ'],
])('getComponentName("%s") = "%s"', (code, name) => {
  expect(getComponentName({ component: code })).toEqual(name);
});


it.each([
  ['04/07/2022', new Date(2022, 6, 4)],
  ['1/2/2023', new Date(2023, 1, 1)],
])('parseBackwardsDateString(%s) = %s', (input, output) => {
  expect(parseBackwardsDateString(input)).toEqual(output);
});


it.each([
  [new Date(2022, 6, 3), new Date(2022, 6, 4)],
  [new Date(2022, 6, 4), new Date(2022, 6, 4)],
  [new Date(2022, 6, 5), new Date(2022, 6, 4)],
  [new Date(2022, 6, 8), new Date(2022, 6, 4)],
  [new Date(2022, 6, 9), new Date(2022, 6, 11)],
  [new Date(2022, 6, 10), new Date(2022, 6, 11)],
])('closestMonday(%s) = %s', (input, output) => {
  expect(closestMonday(input)).toEqual(output);
});
