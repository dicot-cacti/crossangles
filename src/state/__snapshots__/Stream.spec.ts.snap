// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linkStream gives consistent output 1`] = `
Object {
  "component": "LEC",
  "course": Object {
    "code": "RING9731",
    "id": Object {
      "full": "RING9731",
    },
    "name": "Introduction to Ring Theory",
    "streams": Array [
      Object {
        "component": "LEC",
        "enrols": Array [
          50,
          100,
        ],
        "times": Array [
          Object {
            "canClash": true,
            "location": "Morder",
            "time": "M9",
          },
          Object {
            "location": "Helm's Deep",
            "time": "H12",
            "weeks": "1,3,7-9",
          },
          Object {
            "location": "Tol Brandir",
            "time": "T12",
            "weeks": "1,3,7-9",
          },
        ],
      },
      Object {
        "component": "LEC",
        "enrols": Array [
          10,
          50,
        ],
        "times": Array [
          Object {
            "canClash": true,
            "location": "Morder",
            "time": "T9",
          },
          Object {
            "location": "Helm's Deep",
            "time": "F9",
            "weeks": "1,3,7-9",
          },
          Object {
            "location": "Tol Brandir",
            "time": "W12",
            "weeks": "1,3,7-9",
          },
        ],
      },
      Object {
        "component": "TUT",
        "enrols": Array [
          5,
          10,
        ],
        "times": Array [
          Object {
            "location": "Hobbiton",
            "time": "H9",
          },
        ],
      },
      Object {
        "component": "TUT",
        "enrols": Array [
          9,
          9,
        ],
        "times": Array [
          Object {
            "location": "Fangorn",
            "time": "F19",
          },
        ],
      },
    ],
  },
  "enrols": Array [
    50,
    100,
  ],
  "id": "RING9731~LEC~M9,H12,T12",
  "sessions": Array [
    Object {
      "canClash": true,
      "course": Object {
        "code": "RING9731",
        "id": Object {
          "full": "RING9731",
        },
        "name": "Introduction to Ring Theory",
        "streams": Array [
          Object {
            "component": "LEC",
            "enrols": Array [
              50,
              100,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "M9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "H12",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "T12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "LEC",
            "enrols": Array [
              10,
              50,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "T9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "F9",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "W12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              5,
              10,
            ],
            "times": Array [
              Object {
                "location": "Hobbiton",
                "time": "H9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              9,
              9,
            ],
            "times": Array [
              Object {
                "location": "Fangorn",
                "time": "F19",
              },
            ],
          },
        ],
      },
      "day": "M",
      "end": 10,
      "id": "RING9731~LEC~M9,H12,T12~0",
      "index": 0,
      "location": "Morder",
      "start": 9,
      "stream": [Circular],
      "weeks": undefined,
    },
    Object {
      "canClash": undefined,
      "course": Object {
        "code": "RING9731",
        "id": Object {
          "full": "RING9731",
        },
        "name": "Introduction to Ring Theory",
        "streams": Array [
          Object {
            "component": "LEC",
            "enrols": Array [
              50,
              100,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "M9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "H12",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "T12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "LEC",
            "enrols": Array [
              10,
              50,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "T9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "F9",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "W12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              5,
              10,
            ],
            "times": Array [
              Object {
                "location": "Hobbiton",
                "time": "H9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              9,
              9,
            ],
            "times": Array [
              Object {
                "location": "Fangorn",
                "time": "F19",
              },
            ],
          },
        ],
      },
      "day": "H",
      "end": 13,
      "id": "RING9731~LEC~M9,H12,T12~1",
      "index": 1,
      "location": "Helm's Deep",
      "start": 12,
      "stream": [Circular],
      "weeks": "1,3,7-9",
    },
    Object {
      "canClash": undefined,
      "course": Object {
        "code": "RING9731",
        "id": Object {
          "full": "RING9731",
        },
        "name": "Introduction to Ring Theory",
        "streams": Array [
          Object {
            "component": "LEC",
            "enrols": Array [
              50,
              100,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "M9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "H12",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "T12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "LEC",
            "enrols": Array [
              10,
              50,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "T9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "F9",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "W12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              5,
              10,
            ],
            "times": Array [
              Object {
                "location": "Hobbiton",
                "time": "H9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              9,
              9,
            ],
            "times": Array [
              Object {
                "location": "Fangorn",
                "time": "F19",
              },
            ],
          },
        ],
      },
      "day": "T",
      "end": 13,
      "id": "RING9731~LEC~M9,H12,T12~2",
      "index": 2,
      "location": "Tol Brandir",
      "start": 12,
      "stream": [Circular],
      "weeks": "1,3,7-9",
    },
  ],
  "times": Array [
    Object {
      "canClash": true,
      "location": "Morder",
      "time": "M9",
    },
    Object {
      "location": "Helm's Deep",
      "time": "H12",
      "weeks": "1,3,7-9",
    },
    Object {
      "location": "Tol Brandir",
      "time": "T12",
      "weeks": "1,3,7-9",
    },
  ],
}
`;

exports[`linkStream gives consistent output 2`] = `
Object {
  "component": "LEC",
  "course": Object {
    "code": "RING9731",
    "id": Object {
      "full": "RING9731",
    },
    "name": "Introduction to Ring Theory",
    "streams": Array [
      Object {
        "component": "LEC",
        "enrols": Array [
          50,
          100,
        ],
        "times": Array [
          Object {
            "canClash": true,
            "location": "Morder",
            "time": "M9",
          },
          Object {
            "location": "Helm's Deep",
            "time": "H12",
            "weeks": "1,3,7-9",
          },
          Object {
            "location": "Tol Brandir",
            "time": "T12",
            "weeks": "1,3,7-9",
          },
        ],
      },
      Object {
        "component": "LEC",
        "enrols": Array [
          10,
          50,
        ],
        "times": Array [
          Object {
            "canClash": true,
            "location": "Morder",
            "time": "T9",
          },
          Object {
            "location": "Helm's Deep",
            "time": "F9",
            "weeks": "1,3,7-9",
          },
          Object {
            "location": "Tol Brandir",
            "time": "W12",
            "weeks": "1,3,7-9",
          },
        ],
      },
      Object {
        "component": "TUT",
        "enrols": Array [
          5,
          10,
        ],
        "times": Array [
          Object {
            "location": "Hobbiton",
            "time": "H9",
          },
        ],
      },
      Object {
        "component": "TUT",
        "enrols": Array [
          9,
          9,
        ],
        "times": Array [
          Object {
            "location": "Fangorn",
            "time": "F19",
          },
        ],
      },
    ],
  },
  "enrols": Array [
    10,
    50,
  ],
  "id": "RING9731~LEC~T9,F9,W12",
  "sessions": Array [
    Object {
      "canClash": true,
      "course": Object {
        "code": "RING9731",
        "id": Object {
          "full": "RING9731",
        },
        "name": "Introduction to Ring Theory",
        "streams": Array [
          Object {
            "component": "LEC",
            "enrols": Array [
              50,
              100,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "M9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "H12",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "T12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "LEC",
            "enrols": Array [
              10,
              50,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "T9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "F9",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "W12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              5,
              10,
            ],
            "times": Array [
              Object {
                "location": "Hobbiton",
                "time": "H9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              9,
              9,
            ],
            "times": Array [
              Object {
                "location": "Fangorn",
                "time": "F19",
              },
            ],
          },
        ],
      },
      "day": "T",
      "end": 10,
      "id": "RING9731~LEC~T9,F9,W12~0",
      "index": 0,
      "location": "Morder",
      "start": 9,
      "stream": [Circular],
      "weeks": undefined,
    },
    Object {
      "canClash": undefined,
      "course": Object {
        "code": "RING9731",
        "id": Object {
          "full": "RING9731",
        },
        "name": "Introduction to Ring Theory",
        "streams": Array [
          Object {
            "component": "LEC",
            "enrols": Array [
              50,
              100,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "M9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "H12",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "T12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "LEC",
            "enrols": Array [
              10,
              50,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "T9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "F9",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "W12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              5,
              10,
            ],
            "times": Array [
              Object {
                "location": "Hobbiton",
                "time": "H9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              9,
              9,
            ],
            "times": Array [
              Object {
                "location": "Fangorn",
                "time": "F19",
              },
            ],
          },
        ],
      },
      "day": "F",
      "end": 10,
      "id": "RING9731~LEC~T9,F9,W12~1",
      "index": 1,
      "location": "Helm's Deep",
      "start": 9,
      "stream": [Circular],
      "weeks": "1,3,7-9",
    },
    Object {
      "canClash": undefined,
      "course": Object {
        "code": "RING9731",
        "id": Object {
          "full": "RING9731",
        },
        "name": "Introduction to Ring Theory",
        "streams": Array [
          Object {
            "component": "LEC",
            "enrols": Array [
              50,
              100,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "M9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "H12",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "T12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "LEC",
            "enrols": Array [
              10,
              50,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "T9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "F9",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "W12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              5,
              10,
            ],
            "times": Array [
              Object {
                "location": "Hobbiton",
                "time": "H9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              9,
              9,
            ],
            "times": Array [
              Object {
                "location": "Fangorn",
                "time": "F19",
              },
            ],
          },
        ],
      },
      "day": "W",
      "end": 13,
      "id": "RING9731~LEC~T9,F9,W12~2",
      "index": 2,
      "location": "Tol Brandir",
      "start": 12,
      "stream": [Circular],
      "weeks": "1,3,7-9",
    },
  ],
  "times": Array [
    Object {
      "canClash": true,
      "location": "Morder",
      "time": "T9",
    },
    Object {
      "location": "Helm's Deep",
      "time": "F9",
      "weeks": "1,3,7-9",
    },
    Object {
      "location": "Tol Brandir",
      "time": "W12",
      "weeks": "1,3,7-9",
    },
  ],
}
`;

exports[`linkStream gives consistent output 3`] = `
Object {
  "component": "TUT",
  "course": Object {
    "code": "RING9731",
    "id": Object {
      "full": "RING9731",
    },
    "name": "Introduction to Ring Theory",
    "streams": Array [
      Object {
        "component": "LEC",
        "enrols": Array [
          50,
          100,
        ],
        "times": Array [
          Object {
            "canClash": true,
            "location": "Morder",
            "time": "M9",
          },
          Object {
            "location": "Helm's Deep",
            "time": "H12",
            "weeks": "1,3,7-9",
          },
          Object {
            "location": "Tol Brandir",
            "time": "T12",
            "weeks": "1,3,7-9",
          },
        ],
      },
      Object {
        "component": "LEC",
        "enrols": Array [
          10,
          50,
        ],
        "times": Array [
          Object {
            "canClash": true,
            "location": "Morder",
            "time": "T9",
          },
          Object {
            "location": "Helm's Deep",
            "time": "F9",
            "weeks": "1,3,7-9",
          },
          Object {
            "location": "Tol Brandir",
            "time": "W12",
            "weeks": "1,3,7-9",
          },
        ],
      },
      Object {
        "component": "TUT",
        "enrols": Array [
          5,
          10,
        ],
        "times": Array [
          Object {
            "location": "Hobbiton",
            "time": "H9",
          },
        ],
      },
      Object {
        "component": "TUT",
        "enrols": Array [
          9,
          9,
        ],
        "times": Array [
          Object {
            "location": "Fangorn",
            "time": "F19",
          },
        ],
      },
    ],
  },
  "enrols": Array [
    5,
    10,
  ],
  "id": "RING9731~TUT~H9",
  "sessions": Array [
    Object {
      "canClash": undefined,
      "course": Object {
        "code": "RING9731",
        "id": Object {
          "full": "RING9731",
        },
        "name": "Introduction to Ring Theory",
        "streams": Array [
          Object {
            "component": "LEC",
            "enrols": Array [
              50,
              100,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "M9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "H12",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "T12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "LEC",
            "enrols": Array [
              10,
              50,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "T9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "F9",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "W12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              5,
              10,
            ],
            "times": Array [
              Object {
                "location": "Hobbiton",
                "time": "H9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              9,
              9,
            ],
            "times": Array [
              Object {
                "location": "Fangorn",
                "time": "F19",
              },
            ],
          },
        ],
      },
      "day": "H",
      "end": 10,
      "id": "RING9731~TUT~H9~0",
      "index": 0,
      "location": "Hobbiton",
      "start": 9,
      "stream": [Circular],
      "weeks": undefined,
    },
  ],
  "times": Array [
    Object {
      "location": "Hobbiton",
      "time": "H9",
    },
  ],
}
`;

exports[`linkStream gives consistent output 4`] = `
Object {
  "component": "TUT",
  "course": Object {
    "code": "RING9731",
    "id": Object {
      "full": "RING9731",
    },
    "name": "Introduction to Ring Theory",
    "streams": Array [
      Object {
        "component": "LEC",
        "enrols": Array [
          50,
          100,
        ],
        "times": Array [
          Object {
            "canClash": true,
            "location": "Morder",
            "time": "M9",
          },
          Object {
            "location": "Helm's Deep",
            "time": "H12",
            "weeks": "1,3,7-9",
          },
          Object {
            "location": "Tol Brandir",
            "time": "T12",
            "weeks": "1,3,7-9",
          },
        ],
      },
      Object {
        "component": "LEC",
        "enrols": Array [
          10,
          50,
        ],
        "times": Array [
          Object {
            "canClash": true,
            "location": "Morder",
            "time": "T9",
          },
          Object {
            "location": "Helm's Deep",
            "time": "F9",
            "weeks": "1,3,7-9",
          },
          Object {
            "location": "Tol Brandir",
            "time": "W12",
            "weeks": "1,3,7-9",
          },
        ],
      },
      Object {
        "component": "TUT",
        "enrols": Array [
          5,
          10,
        ],
        "times": Array [
          Object {
            "location": "Hobbiton",
            "time": "H9",
          },
        ],
      },
      Object {
        "component": "TUT",
        "enrols": Array [
          9,
          9,
        ],
        "times": Array [
          Object {
            "location": "Fangorn",
            "time": "F19",
          },
        ],
      },
    ],
  },
  "enrols": Array [
    9,
    9,
  ],
  "id": "RING9731~TUT~F19",
  "sessions": Array [
    Object {
      "canClash": undefined,
      "course": Object {
        "code": "RING9731",
        "id": Object {
          "full": "RING9731",
        },
        "name": "Introduction to Ring Theory",
        "streams": Array [
          Object {
            "component": "LEC",
            "enrols": Array [
              50,
              100,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "M9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "H12",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "T12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "LEC",
            "enrols": Array [
              10,
              50,
            ],
            "times": Array [
              Object {
                "canClash": true,
                "location": "Morder",
                "time": "T9",
              },
              Object {
                "location": "Helm's Deep",
                "time": "F9",
                "weeks": "1,3,7-9",
              },
              Object {
                "location": "Tol Brandir",
                "time": "W12",
                "weeks": "1,3,7-9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              5,
              10,
            ],
            "times": Array [
              Object {
                "location": "Hobbiton",
                "time": "H9",
              },
            ],
          },
          Object {
            "component": "TUT",
            "enrols": Array [
              9,
              9,
            ],
            "times": Array [
              Object {
                "location": "Fangorn",
                "time": "F19",
              },
            ],
          },
        ],
      },
      "day": "F",
      "end": 20,
      "id": "RING9731~TUT~F19~0",
      "index": 0,
      "location": "Fangorn",
      "start": 19,
      "stream": [Circular],
      "weeks": undefined,
    },
  ],
  "times": Array [
    Object {
      "location": "Fangorn",
      "time": "F19",
    },
  ],
}
`;
